austation/code/modules/coilgun/charger.dm

```diff
@@ -0,0 +1,152 @@
+	var/obj/structure/disposalpipe/coilgun/charger/C
+	for(var/turf/T in range(1, src)) // for every tile next to the charger
+		C = locate() in T
+		if(C && C.dpdir == P.dpdir && (!C.parent || C.parent == C) && C.laststep != laststep)
+			if(laststep == C)
+				continue
+			laststep = src
+			if(!(C in P.members))
+				P.members += C
+				C.parent = P
+			C.visible_message("<span class='warning'>Debug: building charger...</span>")
+			C.build_charger(P)
+			C.visible_message("<span class='warning'>Debug: synced with parent!</span>")
+
+/obj/structure/disposalpipe/coilgun/charger/proc/check_connection(obj/structure/disposalpipe/coilgun/charger/P)
+	for(var/obj/structure/disposalpipe/coilgun/charger/C in P.members)
+		var/turf/T = loc
+		if(isturf(T) && !T.intact)
+			attached = locate() in T
+			if(attached)
+				return TRUE
+
+	return FALSE
+
+/obj/structure/disposalpipe/coilgun/charger/proc/power_process()
+	if(current_power_use && target_power_usage)
+		var/datum/powernet/PN = attached.powernet
+		if(attached && PN)
+			var/drained = min(current_power_use / (target_power_usage / 100), max_power_use)// set our power use
+			visible_message("<span class='warning'>Drained reads [drained]!</span>") // DEBUG
+			attached.add_delayedload(drained) // apply our power use to the connected wire
+			if(attached.newavail() < drained) // are we using more power than we have connected?
+				target_power_usage -= 20 // throttle it down
+				if(target_power_usage > 0)
+					visible_message("<span class='warning'>\The [src]'s power warning light flickers, lowering throttle to [target_power_usage]!</span>")
+					return power_process() // check if the reduced power load is enough
+			else
+				current_power_use = drained
+				return TRUE
+	enabled = FALSE // if we failed any of the other checks, disable the charger
+	visible_message("<span class='warning'>\The [src]'s power warning light flickers, turning itself off!</span>")
+	return FALSE
+
+/obj/structure/disposalpipe/coilgun/charger/transfer(obj/structure/disposalholder/H)
+	if(H.contents.len)
+		if(enabled && power_process()) // is this enabled, do we have enough power?
+			for(var/atom/movable/AM in H.contents) // run the loop below for every movable that passes through the charger
+				if(istype(AM, /obj/effect/hvp)) // if it's a coilgun projectile, continue
+
+					var/obj/effect/hvp/PJ = AM
+					var/datum/powernet/PN = attached.powernet
+
+					if(PN && target_power_usage)
+						var/prelim = (current_power_use + POWER_DIVIDER) / POWER_DIVIDER
+						visible_message("<span class='danger'>debug: prelim reads [prelim]!</span>") // DEBUG
+//						current_power_use = min((P.p_speed * 500) * (target_power_usage / 100), max_power_use) //determins power usage
+						speed_increase = prelim * BASE ** PJ.p_speed
+						PJ.p_speed += speed_increase * members.len
+						PJ.p_heat += heat_increase * members.len
```

Comment:
`* members.len` is an interesting choice.
Is it not going to be processed by every other coilgun piece in the gun?
I thought the speed gain would be incremental with each piece it passed through rather than all at once

Reply:
member and linkage is debug